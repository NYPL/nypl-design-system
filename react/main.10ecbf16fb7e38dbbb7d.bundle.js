(window.webpackJsonp=window.webpackJsonp||[]).push([[0],{1292:function(n,o,c){"use strict";c.r(o);c(1293),c(1323),c(1335),c(1442),c(1485)},1448:function(n,o){},1449:function(n,o){},1450:function(n,o){},1451:function(n,o){},501:function(n,o,c){c(502),c(645),n.exports=c(1292)},564:function(n,o){}},[[501,1,2]]]);