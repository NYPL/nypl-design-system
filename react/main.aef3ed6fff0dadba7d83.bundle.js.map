{"version":3,"file":"main.aef3ed6fff0dadba7d83.bundle.js","sources":["webpack:///main.aef3ed6fff0dadba7d83.bundle.js"],"mappings":"AAAA","sourceRoot":""}