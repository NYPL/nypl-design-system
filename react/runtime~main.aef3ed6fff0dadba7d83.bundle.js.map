{"version":3,"file":"runtime~main.aef3ed6fff0dadba7d83.bundle.js","sources":["webpack:///runtime~main.aef3ed6fff0dadba7d83.bundle.js"],"mappings":"AAAA","sourceRoot":""}