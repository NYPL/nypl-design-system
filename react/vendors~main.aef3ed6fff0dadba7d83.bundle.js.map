{"version":3,"file":"vendors~main.aef3ed6fff0dadba7d83.bundle.js","sources":["webpack:///vendors~main.aef3ed6fff0dadba7d83.bundle.js"],"mappings":";AAAA","sourceRoot":""}