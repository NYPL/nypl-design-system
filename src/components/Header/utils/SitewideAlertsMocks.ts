// This is based on the Refinery API endpoint response.
const getApiResponse = (
  startDate = "2022-03-14T00:00:00-04:00",
  endDate = "2022-12-31T23:59:00-05:00"
) => ({
  data: [
    {
      attributes: {
        "alert-text": {
          en: {
            text: '<p>Masks are encouraged at all NYPL locations. <a href="https://nypl.org/reopening" rel="nofollow">Learn more and find information about our hours and services.</a></p>\n',
            type: "text-single",
          },
          type: "text-group",
        },
        "closed-reason": {
          en: {
            text: '<p>Masks are encouraged at all NYPL locations. <a href="https://nypl.org/reopening" rel="nofollow">Learn more and find information about our hours and services.</a></p>\n',
            type: "text-single",
          },
          type: "text-group",
        },
        "closing-date-end": null,
        "closing-date-start": null,
        "date-created": null,
        "date-modified": null,
        "display-date-end": endDate,
        "display-date-start": startDate,
        extended: null,
        locations: null,
        scope: "all",
        uri: { type: "uri", "full-uri": "", description: "" },
        uuid: null,
      },
      id: "966",
      links: {
        self: "https://refinery.nypl.org/api/nypl/ndo/v0.1/content/alerts/966",
      },
      type: "alert",
    },
  ],
  jsonapi: { version: "1.0" },
  links: { self: "https://refinery.nypl.org/api/nypl/ndo/v0.1/content/alerts" },
  meta: { count: 1 },
});
export const refineryResponse = getApiResponse();
export const expiredAlertsReponse = getApiResponse(
  "2020-03-14T00:00:00-04:00",
  "2020-12-31T23:59:00-05:00"
);
