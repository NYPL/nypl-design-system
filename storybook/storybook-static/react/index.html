<!doctype html><html lang="en"><head><meta charset="utf-8"><title>Storybook</title><meta name="viewport" content="width=device-width,initial-scale=1"><style>html, body {
    overflow: hidden;
    height: 100%;
    width: 100%;
    margin: 0;
    padding: 0;
  }</style><script>/* globals window */
  /* eslint-disable no-underscore-dangle */
  try {
    if (window.top !== window) {
      window.__REACT_DEVTOOLS_GLOBAL_HOOK__ = window.top.__REACT_DEVTOOLS_GLOBAL_HOOK__;
    }
  } catch (e) {
    // eslint-disable-next-line no-console
    console.warn('unable to connect to top frame for connecting dev tools');
  }</script><style>#root[hidden],
      #docs-root[hidden] {
        display: none !important;
      }</style></head><body><div id="root"></div><div id="docs-root"></div><script>window['DOCS_MODE'] = false;</script><script src="runtime~main.00bd9d033d7a0b972fc9.bundle.js"></script><script src="vendors~main.d2b6aafef1a259d165c0.bundle.js"></script><script src="main.3bf71672a6dd35499004.bundle.js"></script></body></html>