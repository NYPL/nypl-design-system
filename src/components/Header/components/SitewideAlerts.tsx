import { chakra, useStyleConfig } from "@chakra-ui/react";
import React, { useState, useEffect } from "react";

import List from "../../List/List";
import Notification from "../../Notification/Notification";
import { Alert, alertsApiUrl, parseAlertsData } from "../headerUtils";

/**
 * The SitewideAlerts component makes an API request to an NYPL API endpoint
 * to fetch NYPL sitewide alerts. While this component can be used in
 * isolation, it is already rendered in the DS Header component.
 */
export const SitewideAlerts = chakra(({ isMobile }) => {
  const [alerts, setAlerts] = useState<Alert[]>([]);
  const styles = useStyleConfig("SitewideAlerts", { isMobile });
  const fetchErrorMessage =
    "NYPL Reservoir SitewideAlerts: There was an error fetching NYPL sitewide alerts.";

  // Make a request to the NYPL API endpoint for sitewide alerts, parse
  // the data, and set it to the local state.
  useEffect(() => {
    fetch(alertsApiUrl)
      .then((response) => {
        if (response.status >= 200 && response.status < 300) {
          return response.json();
        } else {
          throw new Error(fetchErrorMessage);
        }
      })
      .then((data) => {
        const parsedData: Alert[] = parseAlertsData(data);
        setAlerts(parsedData);
      })
      .catch((_error) => {
        // Override any error message with our own.
        console.warn(fetchErrorMessage);
      });
  }, []);

  // Great, we have NYPL alerts data. Each alert is rendered as a list
  // item in an unordered list. Note that the alert description may
  // contain HTML which is rendered as-is.
  const getAlertsElems = (data: Alert[]) => {
    return (
      <List noStyling type="ul">
        {data.map((alert: Alert) => (
          <li
            key={alert.id}
            dangerouslySetInnerHTML={{ __html: alert.description }}
          />
        ))}
      </List>
    );
  };

  return alerts.length > 0 ? (
    <Notification
      aria-label="Sitewide alerts"
      id="sitewideAlerts-notification"
      isCentered
      noMargin
      notificationContent={getAlertsElems(alerts)}
      showIcon={false}
      __css={styles}
    />
  ) : null;
});

export default SitewideAlerts;
