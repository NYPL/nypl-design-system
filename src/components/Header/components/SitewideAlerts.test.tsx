import { render, screen, waitFor, within } from "@testing-library/react";
import { axe } from "jest-axe";
import * as React from "react";
import renderer from "react-test-renderer";

import SitewideAlerts from "./SitewideAlerts";
import { expiredAlertsReponse, refineryResponse } from "./SitewideAlertsMocks";

describe("SitewideAlerts Accessibility", () => {
  it("passes axe accessibility test", async () => {
    (global as any).fetch = jest.fn(() =>
      Promise.resolve({
        status: 200,
        json: () => Promise.resolve(refineryResponse),
      })
    ) as jest.Mock;

    const { container } = await waitFor(() => render(<SitewideAlerts />));
    expect(await axe(container)).toHaveNoViolations();
  });
});

describe("SitewideAlerts", () => {
  describe("Successful API request", () => {
    beforeEach(async () => {
      (global as any).fetch = jest.fn(() =>
        Promise.resolve({
          status: 200,
          json: () => Promise.resolve(refineryResponse),
        })
      ) as jest.Mock;

      await waitFor(() => render(<SitewideAlerts />));
    });

    it("renders a notification", () => {
      const notification = screen.getByRole("complementary");

      expect(notification).toHaveAttribute("aria-label", "Sitewide alerts");
      expect(notification).toHaveAttribute("data-type", "standard");
    });

    it("renders a list with text", () => {
      const notification = screen.getByRole("complementary");
      const { getByText, getAllByRole } = within(notification);
      const list = getAllByRole("list");

      expect(list.length).toBe(1);
      expect(getByText(/masks are encouraged/i)).toBeInTheDocument();
    });

    it("renders the UI snapshot correctly", () => {
      const sitewideAlerts = renderer.create(<SitewideAlerts />).toJSON();

      expect(sitewideAlerts).toMatchSnapshot();
    });
  });

  // Since alerts are expired, the component should not render.
  describe("Successful API request but expired alerts", () => {
    it("renders a notification", async () => {
      (global as any).fetch = jest.fn(() =>
        Promise.resolve({
          status: 200,
          json: () => Promise.resolve(expiredAlertsReponse),
        })
      ) as jest.Mock;
      await waitFor(() => render(<SitewideAlerts />));

      expect(screen.queryByRole("complementary")).not.toBeInTheDocument();
    });
  });

  // If there was an API request error, the component should not render.
  describe("Unsuccessful API request", () => {
    it("does not render HTML", async () => {
      const warn = jest.spyOn(console, "warn");
      (global as any).fetch = jest.fn(() =>
        Promise.reject(new Error("API request failed"))
      ) as jest.Mock;
      await waitFor(() => render(<SitewideAlerts />));

      expect(screen.queryByRole("complementary")).not.toBeInTheDocument();
      expect(warn).toHaveBeenCalledWith(
        "NYPL Reservoir SitewideAlerts: There was an error fetching NYPL sitewide alerts."
      );
    });
  });
});
