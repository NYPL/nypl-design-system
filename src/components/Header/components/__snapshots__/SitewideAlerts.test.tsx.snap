// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SitewideAlerts Successful API request renders the UI snapshot correctly 1`] = `null`;
